### PR TITLE
feat(manager:npm): support `binarySource=install` for yarn

### LINKED ISSUE #14797: Yarn lock file maintenance using wrong version
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When Yarn 2+ is used in a project, but not embedded in the repository via `yarnPath`, lock file maintenance will fall back to Yarn 1 and replace the whole `yarn.lock`.

An example can be seen here: https://github.com/jgraichen/renovate-corepack-yarn/pull/4/files

When dependencies are bumped, the `yarn.lock` is correctly updated: https://github.com/jgraichen/renovate-corepack-yarn/pull/3/files

I'd love to have all updates and lock file maintenance use the correct Yarn version.

### Relevant debug logs

_No response_

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by viceice**
## Changes
- support `binarySource=install` for yarn
- use `corepack` if `packageManager` field is found.

## Context
- closes #14797
- closes #15173

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/yarn.ts [hunk 1]
```diff
@@ -277,3 +292,11 @@ export async function generateLockFile(
...
   return { lockFile };
 }
+async function isCorepack(lockFileDir: string): Promise<boolean> {
+  const pkgJson = await readLocalFile(
+    upath.join(lockFileDir, 'package.json'),
+    'utf8'
+  );
+
+  return is.nonEmptyStringAndNotWhitespace(JSON.parse(pkgJson)?.packageManager);
+}
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/yarn.ts:302]: @rarkins I'm thinking if it makes sense to do this at extrsaction time and add to manager config object, so we don't need to read the package file again 🤔

**Comment by viceice** [comment]
Don't merge yet, needs https://github.com/containerbase/node/runs/6270509785?check_suite_focus=true to finish

File: lib/modules/manager/npm/post-update/yarn.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/manager/npm/post-update/yarn.ts:302]: Yeah, anything which doesn't change could be done at either stage

Comment (rarkins) [reply lib/modules/manager/npm/post-update/yarn.ts:302]: we don't read packagel.json for other reasons though in post-update? I thought we did

Comment (viceice) [reply lib/modules/manager/npm/post-update/yarn.ts]: nope, we don't 🤷‍♂️ 

fixed by moving to extraction phase.

File: lib/modules/manager/npm/post-update/__snapshots__/yarn.spec.ts.snap [hunk 2]
```diff
@@ -365,6 +365,49 @@ Array [
 ]
 `;
 
+exports[`modules/manager/npm/post-update/yarn patchest local yarn 1`] = `
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/__snapshots__/yarn.spec.ts.snap:368]:
```suggestion
exports[`modules/manager/npm/post-update/yarn patches local yarn 1`] = `
```

File: lib/modules/manager/npm/post-update/yarn.ts [hunk 3]
```diff
@@ -103,12 +109,21 @@ export async function generateLockFile(
...
-    }
+    const preCommands: string[] = [];
+
+    const yarnTool: ToolConstraint = {
+      toolName: 'yarn',
+      constraint: '^1.22.18', // needs to be a v1 yarn, otherwise v2 will be installed
+    };
 
-    const preCommands = [installYarn];
+    if (config.managerData?.hasPackageManager) {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/yarn.ts:119]: @rarkins should we fallback to yarn v1 install when `isYarn1===true`, so we can use the `slim` (patched) yarn?

**Comment by viceice** [comment]
needs another buble
- https://github.com/containerbase/buildpack/commit/1307333e8ff2adb7339422c9bc0830013a8aca9f

**Comment by viceice** [comment]
works 🥳 🎉 

https://github.com/renovate-reproductions/14797-renovate-corepack-yarn/pulls

File: lib/modules/manager/npm/post-update/yarn.ts (hunk 3 above)
Comment (rarkins) [reply lib/modules/manager/npm/post-update/yarn.ts:119]: Yes, that would be very desirable. It saves a lot of time and bandwidth

**Review by rarkins — APPROVED** [review]

**Comment by bobvandevijver** [comment]
@viceice Since this morning I'm seeing `bash: yarn: command not found` in my newly created renovate mrs, could this be related to this change?

**Comment by rarkins** [comment]
Created #15458 to prepare for possible revert of recent yarn changes

**Comment by rarkins** [comment]
@bobvandevijver hosted app or self-hosted?

**Comment by bobvandevijver** [comment]
Self-hosted on Gitlab, using the slim image version v32.41.2.

**Comment by rarkins** [comment]
Can you copy paste here the `docker run` command for one of the commands which failed? We think it's related to tags with a `v` prefix, e.g. selecting `v16.10.0` instead of `16.10.0`

**Comment by bobvandevijver** [comment]
```
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","image":"node","msg":"Using docker to execute","time":"2022-05-05T07:30:09.380Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","toolName":"yarn-slim","constraint":"^1.22.18","resolvedVersion":"1.22.18","msg":"Resolved version","time":"2022-05-05T07:30:09.547Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","image":"docker.io/renovate/node","msg":"No tag or tagConstraint specified","time":"2022-05-05T07:30:09.547Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","msg":"Fetching Docker image: docker.io/renovate/node","time":"2022-05-05T07:30:09.548Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","msg":"Finished fetching Docker image","time":"2022-05-05T07:30:10.782Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","command":"docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"","msg":"Executing command","time":"2022-05-05T07:30:10.836Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","err":{"code":127,"killed":false,"signal":null,"cmd":"docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"","stdout":"Installing legacy tool yarn-slim v1.22.18\n\nadded 1 package in 1s\n1.22.18\n","stderr":"bash: yarn: command not found\n","message":"Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nbash: yarn: command not found\n","stack":"Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nbash: yarn: command not found\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"msg":"rawExec err","time":"2022-05-05T07:30:13.833Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":14,"level":20,"logContext":"2fGWC0V2gIC-U8Ca6fG5F","repository":"extern/project","branch":"renovate/core-js-3.x","err":{"code":127,"killed":false,"signal":null,"cmd":"docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"","stdout":"Installing legacy tool yarn-slim v1.22.18\n\nadded 1 package in 1s\n1.22.18\n","stderr":"bash: yarn: command not found\n","message":"Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nbash: yarn: command not found\n","stack":"Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/cache/renovate/drenso/repos/gitlab/extern/project\":\"/cache/renovate/drenso/repos/gitlab/extern/project\" -v \"/cache/renovate/drenso/cache\":\"/cache/renovate/drenso/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_CACHE_FOLDER -w \"/cache/renovate/drenso/repos/gitlab/extern/project\" docker.io/renovate/node bash -l -c \"install-tool yarn-slim 1.22.18 && yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\"\nbash: yarn: command not found\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"type":"yarn","msg":"lock file error","time":"2022-05-05T07:30:13.896Z","v":0}
```

**Comment by rarkins** [comment]
Thanks, working on a fix in next 1-2 hours (it requires a lot of images to be rebuilt and pushed to Docker Hub)

**Comment by bobvandevijver** [comment]
Yes, it is fixed now 😃

**Pull request merged**